### PR TITLE
Jackson 1 instrumentation for IAST

### PR DESCRIPTION
**Pull request by DDJavierSantos**
# What Does This Do

Instrument the older versions of the jackson library.

# Motivation
org.codehaus.jackson is used by Jersey 2 to deserialize json objects. We need to instrument the parser calls so we can achieve propagation in IAST.

# Additional Notes

### CODE REVIEW

File: dd-java-agent/instrumentation/glassfish/build.gradle [hunk 1]
```diff
@@ -27,6 +27,8 @@ dependencies {
   testImplementation project(':dd-java-agent:instrumentation:grizzly-2')
   testImplementation deps.guava
   testImplementation group: 'org.glassfish.main.extras', name: 'glassfish-embedded-all', version: '4.0'
+  testImplementation group: 'jakarta.ws.rs', name: 'jakarta.ws.rs-api', version: '3.0.0'
```
Comment (smola) [inline dd-java-agent/instrumentation/glassfish/build.gradle:30]: I don't think this is related to this PR?

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderAdvice.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+      final WebModule module = InstrumentationBridge.WEB;
+
+      if (module != null) {
+        try {
+          module.onParameterValue(null, (String) result);
+        } catch (final Throwable e) {
+          module.onUnexpectedException("afterGetParameter threw", e);
+        }
+      } else {
+        System.out.println("Module is null");
```
Comment (smola) [inline dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderAdvice.java:20]: No stdout printing.

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package datadog.trace.instrumentation.jersey3;
+
+import static datadog.trace.agent.tooling.bytebuddy.matcher.NameMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+
+@AutoService(Instrumenter.class)
+public class AbstractStringReaderInstrumentation extends Instrumenter.Tracing
```
Comment (smola) [inline dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderInstrumentation.java:10]:
```suggestion
public class AbstractStringReaderInstrumentation extends Instrumenter.Iast
```

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/instrumentation/glassfish/build.gradle (hunk 1 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/glassfish/build.gradle]: removed

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderInstrumentation.java (hunk 3 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderInstrumentation.java]: fixed

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderAdvice.java (hunk 2 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/AbstractStringReaderAdvice.java]: fixed

File: dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonFactoryCallSite.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  public static JsonParser afterCreate(
+      @CallSite.This @Nonnull final JsonFactory self,
+      @CallSite.Argument @Nullable final Object input,
+      @CallSite.Return @Nullable final JsonParser jsonParser) {
+    final PropagationModule module = InstrumentationBridge.PROPAGATION;
+    if (module != null) {
+      try {
+        module.taintIfInputIsTainted(jsonParser, input);
+      } catch (final Throwable e) {
+        module.onUnexpectedException("afterCreate threw", e);
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonFactoryCallSite.java:26]:
```suggestion
        module.onUnexpectedException("JsonFactoryCallSite.afterCreate threw", e);
```

File: dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonFactoryCallSite.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  public static JsonParser afterCreate(
+      @CallSite.This @Nonnull final JsonFactory self,
+      @CallSite.Argument @Nullable final String content,
+      @CallSite.Return @Nullable final JsonParser jsonParser) {
+    final PropagationModule module = InstrumentationBridge.PROPAGATION;
+    if (module != null) {
+      try {
+        module.taintIfInputIsTainted(jsonParser, content);
+      } catch (final Throwable e) {
+        module.onUnexpectedException("afterCreate threw", e);
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonFactoryCallSite.java:43]:
```suggestion
        module.onUnexpectedException("JsonFactoryCallSite.afterCreate threw", e);
```

File: dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonParserInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.codehaus.jackson.JsonParser;
+
+@AutoService(Instrumenter.class)
+public class JsonParserInstrumentation extends Instrumenter.Iast
+    implements Instrumenter.ForTypeHierarchy {
+
+  public JsonParserInstrumentation() {
+    super("jsonParser");
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/JsonParserInstrumentation.java:28]:
```suggestion
    super("jackson", "jackson-1");
```

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/instrumentation/iast-instrumenter/src/main/resources/datadog/trace/instrumentation/iastinstrumenter/iast_exclusion.trie [hunk 7]
```diff
@@ -132,6 +132,7 @@
 1 org.bson.*
 1 org.clojure.*
 1 org.codehaus.*
+0 org.codehaus.jackson.*
```
Comment (smola) [inline dd-java-agent/instrumentation/iast-instrumenter/src/main/resources/datadog/trace/instrumentation/iastinstrumenter/iast_exclusion.trie:135]: Given that we moved to target instrumentation, we can probably remove this?

File: dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/InputStreamInstrumentation.java [hunk 8]
```diff
@@ -38,7 +37,7 @@ public ElementMatcher<TypeDescription> hierarchyMatcher() {
   @Override
   public void adviceTransformations(AdviceTransformation transformation) {
     transformation.applyAdvice(
-        isConstructor().and(isPublic()).and(takesArgument(0, InputStream.class)),
+        isConstructor().and(takesArgument(0, InputStream.class)),
```
Comment (smola) [inline dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/InputStreamInstrumentation.java:40]: Do we need to instrument private constructors here?

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      final WebModule module = InstrumentationBridge.WEB;
+
+      if (module != null) {
+        try {
+          module.onGetInputStream(inputStream);
+        } catch (final Throwable e) {
+          module.onUnexpectedException("afterGetInputStream threw", e);
+        }
+      } else {
+        System.out.println("Module is null");
```
Comment (smola) [inline dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java:45]: Please, no println. You can remove the `else` altogether.

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/instrumentation/iast-instrumenter/src/main/resources/datadog/trace/instrumentation/iastinstrumenter/iast_exclusion.trie (hunk 7 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/iast-instrumenter/src/main/resources/datadog/trace/instrumentation/iastinstrumenter/iast_exclusion.trie:135]: Done

File: dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/InputStreamInstrumentation.java (hunk 8 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/java-io/src/main/java/datadog/trace/instrumentation/java/lang/InputStreamInstrumentation.java:40]: We need to instrument some protected ones

File: dd-java-agent/agent-iast/src/main/java/com/datadog/iast/source/WebModuleImpl.java [hunk 10]
```diff
@@ -120,4 +121,17 @@ public void onRequestMatrixParameter(
     taintedObjects.taintInputString(
         value, new Source(SourceTypes.REQUEST_MATRIX_PARAMETER, paramName, value));
   }
+
+  @Override
+  public void onGetInputStream(@Nullable InputStream inputStream) {
```
Comment (manuel-alvarez-alvarez) [inline dd-java-agent/agent-iast/src/main/java/com/datadog/iast/source/WebModuleImpl.java:126]:
```
public static ServletInputStream afterGetInputStream(
      @CallSite.This final ServletRequest self,
      @CallSite.Return final ServletInputStream inputStream) {
    final PropagationModule module = InstrumentationBridge.PROPAGATION;
    if (module != null) {
      try {
        module.taint(SourceTypes.REQUEST_BODY, inputStream);
      } catch (final Throwable e) {
        module.onUnexpectedException("afterGetInputStream threw", e);
      }
    }
    return inputStream;
  }
```

might be better as it takes care of the `Taintable` interface

Comment (DDJavierSantos) [reply dd-java-agent/agent-iast/src/main/java/com/datadog/iast/source/WebModuleImpl.java]: applied

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/build.gradle [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+}
+
+apply from: "$rootDir/gradle/java.gradle"
+
+addTestSuiteForDir('latestDepTest', 'test')
+
+dependencies {
+  compileOnly group: 'org.codehaus.jackson', name: 'jackson-core-asl', version: '1.9.13'
+  testImplementation group: 'org.codehaus.jackson', name: 'jackson-core-asl', version: '1.9.13'
+  testRuntimeOnly(project(':dd-java-agent:instrumentation:iast-instrumenter'))
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core/jackson-core-1/build.gradle:14]: Dependency probably not needed if there's no call site in the package.

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return declaresMethod(namedOneOf("getCurrentName", "getText", "nextTextValue"))
+        .and(
+            extendsClass(named(hierarchyMarkerType()))
+                .and(namedNoneOf("org.codehaus.jackson.impl.JsonParserMinimalBase")));
+  }
+
+  public static class JsonParserAdvice {
+
+    @Advice.OnMethodExit
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java:57]: Same here.

File: dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2FactoryTestSuite.java [hunk 13]
```diff
@@ -6,12 +6,12 @@
...
 import java.io.Reader;
 
-public class JsonFactoryTestSuite {
+public class Json2FactoryTestSuite {
 
   private JsonFactory jsonFactory;
 
-  public JsonFactoryTestSuite(final JsonFactory jsonFactory) {
-    this.jsonFactory = jsonFactory;
+  public Json2FactoryTestSuite() {
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2FactoryTestSuite.java:13]: As above, not needed without callsites.

File: dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2ParserTestSuite.java [hunk 14]
```diff
@@ -4,11 +4,11 @@
 import com.fasterxml.jackson.core.JsonToken;
 import java.io.IOException;
 
-public class JsonParserTestSuite {
+public class Json2ParserTestSuite {
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2ParserTestSuite.java:7]: Same here.

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+import datadog.trace.api.iast.propagation.PropagationModule;
+import java.io.InputStream;
+import net.bytebuddy.asm.Advice;
+
+// keep in sync with jersey2 (javax packages)
+@AutoService(Instrumenter.class)
+public class ReaderInterceptorExecutor extends Instrumenter.AppSec
+    implements Instrumenter.ForSingleType {
+  public ReaderInterceptorExecutor() {
+    super("jersey ReaderInterceptorExecutor");
```
Comment (smola) [inline dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java:19]: Please, use a name more in line with the other instrumenters here.

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+
+  @Override
+  public void adviceTransformations(AdviceTransformation transformation) {
+    transformation.applyAdvice(
+        named("getInputStream").and(takesArguments(0)),
+        getClass().getName() + "$ReaderInterceptorExecutorGetInputStreamAdvice");
+  }
+
+  public static class ReaderInterceptorExecutorGetInputStreamAdvice {
+    @Advice.OnMethodExit
```
Comment (smola) [inline dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java:35]: Let's use suppress.

**Review by smola — CHANGES_REQUESTED** [review]

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/build.gradle (hunk 11 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/build.gradle]: changed

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1FactoryInstrumentation.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  @Override
+  public String instrumentedType() {
+    return "org.codehaus.jackson.JsonFactory";
+  }
+
+  public static class InstrumenterAdvice {
+
+    @Advice.OnMethodExit
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1FactoryInstrumentation.java]: applied

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+  }
+
+  @Override
+  public String hierarchyMarkerType() {
+    return "org.codehaus.jackson.JsonParser";
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return declaresMethod(namedOneOf("getCurrentName", "getText", "nextTextValue"))
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java]: applied

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/groovy/Json1FactoryInstrumentationTest.groovy [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+    InstrumentationBridge.registerIastModule(propagationModule)
+    final jsonFactoryTestSuite = new Json1FactoryTestSuite()
+    final content = '{"key":"value"}'
+
+    when:
+    final result = jsonFactoryTestSuite.createParser(content)
+
+    then:
+    result != null
+    1 * propagationModule.taintIfInputIsTainted(_ as JsonParser, content)
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/groovy/Json1FactoryInstrumentationTest.groovy:26]: applied

File: dd-java-agent/instrumentation/jackson-core/src/test/groovy/Json2FactoryInstrumentationTest.groovy [hunk 20]
```diff
@@ -1,24 +1,22 @@
...
 import datadog.trace.agent.test.AgentTestRunner
 import datadog.trace.api.iast.InstrumentationBridge
 import datadog.trace.api.iast.propagation.PropagationModule
-import foo.bar.JsonFactoryTestSuite
-import groovy.transform.CompileDynamic
+import foo.bar.Json2FactoryTestSuite
+
+class Json2FactoryInstrumentationTest extends AgentTestRunner {
 
-@CompileDynamic
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/src/test/groovy/Json2FactoryInstrumentationTest.groovy:9]: reinstated the annotation

File: dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2ParserTestSuite.java (hunk 14 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/src/test/java/foo/bar/Json2ParserTestSuite.java]: removed

File: dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java (hunk 16 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/ReaderInterceptorExecutor.java:35]: applied

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/groovy/Json1FactoryInstrumentationTest.groovy [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    InstrumentationBridge.registerIastModule(propagationModule)
+    final jsonFactoryTestSuite = new Json1FactoryTestSuite()
+    final is = Mock(InputStream)
+
+    when:
+    final result = jsonFactoryTestSuite.createParser(is)
+
+    then:
+    result != null
+    1 * propagationModule.taintIfInputIsTainted(_ as JsonParser, is)
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/groovy/Json1FactoryInstrumentationTest.groovy:41]: applied

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/java/foo/bar/codehaus/Json1FactoryTestSuite.java [hunk 22]
```diff
@@ -0,0 +1,23 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import org.codehaus.jackson.JsonFactory;
+import org.codehaus.jackson.JsonParser;
+
+public class Json1FactoryTestSuite {
+
+  private JsonFactory jsonFactory;
+
+  public Json1FactoryTestSuite() {
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/test/java/foo/bar/codehaus/Json1FactoryTestSuite.java]: removed

File: dd-java-agent/instrumentation/jackson-core/src/main/java/datadog/trace/instrumentation/jackson/core/Json2FactoryInstrumentation.java [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  @Override
+  public String instrumentedType() {
+    return "com.fasterxml.jackson.core.JsonFactory";
+  }
+
+  public static class InstrumenterAdvice {
+
+    @Advice.OnMethodExit
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/src/main/java/datadog/trace/instrumentation/jackson/core/Json2FactoryInstrumentation.java]: applied

File: dd-java-agent/instrumentation/jackson-core/src/test/groovy/Json2FactoryInstrumentationTest.groovy [hunk 24]
```diff
@@ -1,24 +1,22 @@
...
-class JsonFactoryInstrumentationTest extends AgentTestRunner {
 
   @Override
   protected void configurePreAgent() {
-    injectSysConfig('dd.iast.enabled', 'true')
+    injectSysConfig("dd.iast.enabled", "true")
   }
 
-  void 'test createParser(String)'() {
+  def 'test createParser(String)'() {
```
[the remark this answers is not in the document]
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/src/test/groovy/Json2FactoryInstrumentationTest.groovy]: done

File: dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java (hunk 12 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/jackson-core-1/src/main/java/datadog/trace/instrumentation/jackson/codehouse/core/Json1ParserInstrumentation.java]: applied

File: dd-java-agent/instrumentation/jackson-core/build.gradle [hunk 25]
```diff
@@ -16,11 +16,6 @@ muzzle {
 }
 
 apply from: "$rootDir/gradle/java.gradle"
-apply plugin: 'call-site-instrumentation'
```
Comment (smola) [inline dd-java-agent/instrumentation/jackson-core/build.gradle:19]: Below:

```
testRuntimeOnly(project(':dd-java-agent:instrumentation:iast-instrumenter'))
```

I think that shouldn't be needed anymore?

**Review by smola — COMMENTED** [review]
Looks good to me. Just one comment about iast-instrumenter dependency.

File: dd-java-agent/instrumentation/jackson-core/build.gradle (hunk 25 above)
Comment (DDJavierSantos) [reply dd-java-agent/instrumentation/jackson-core/build.gradle:19]: removed

**Review by smola — APPROVED** [review]

**Pull request merged**